library/alloc/src/collections/vec_deque/mod.rs

```diff
@@ -105,13 +91,8 @@ pub struct VecDeque<
     T,
     #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator = Global,
 > {
-    // tail and head are pointers into the buffer. Tail always points
-    // to the first element that could be read, Head always points
-    // to where data should be written.
-    // If tail == head the buffer is empty. The length of the ringbuffer
-    // is defined as the distance between the two.
-    tail: usize,
     head: usize,
```

Comment:
Please include some basic invariant `//` comments here to help someone new to looking at the implementation.

Maybe something like

```rust
    // `self[0]`, if it exists, is `buf[head]`.
    // `head < buf.capacity()`, unless `buf.capacity() == 0` when `head == 0`.
    head: usize,
    // the number of initialized elements, starting from the one at `head` and potentially wrapping around.
    // if `len == 0`, the exact value of `head` is unimportant.
    // because we want to internally use indexes in the range `[0, len * 2 - 1)`,
    // it's not allowed to ever have `len > (1 << (usize::BITS - 2))`.
    len: usize,
```

(And helpfully, you commenting on those comments I wrote will catch if I misunderstand something about how this works.)

Reply:
`len <= isize::MAX as usize` must only hold if `!T::IS_ZST`. I've put great care into making sure that this special case can't lead to incorrect behavior due to integer overflow. To be fair, the fact that it wasn't obvious to you is probably a sign that this kind of comment would be useful.